src/Microsoft.AspNet.TestHost/WebSocketClient.cs

```diff
@@ -0,0 +1,177 @@
+            {
+                pathBase = new PathString(pathBase.Value.Substring(0, pathBase.Value.Length - 1));
+            }
+            _pathBase = pathBase;
+
+            SubProtocols = new List<string>();
+        }
+
+        public IList<string> SubProtocols
+        {
+            get;
+            private set;
+        }
+
+        public Action<HttpRequestMessage> ConfigureRequest
+        {
+            get;
+            set;
+        }
+
+        public async Task<System.Net.WebSockets.WebSocket> ConnectAsync(Uri uri, CancellationToken cancellationToken)
+        {
+            // clientRequest
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, uri);
+            request.Headers.Add("Connection", "Upgrade");
+            request.Headers.Add("Upgrade", "websocket");
+            request.Headers.Add("Sec-WebSocket-Version", "13");
+            request.Headers.Add("Sec-WebSocket-Key", CreateRequestKey());
+            if (SubProtocols.Count > 0)
+            {
+                request.Headers.Add("SecWebSocketProtocol", string.Join(", ", SubProtocols));
+            }
+            if (ConfigureRequest != null)
+            {
+                ConfigureRequest(request);
+            }
+
+            var state = new RequestState(request, _pathBase, cancellationToken);
+
+            // Async offload, don't let the test code block the caller.
+            var offload = Task.Factory.StartNew(async () =>
+            {
+                try
+                {
+                    await _next(state.FeatureCollection);
+                    state.PipelineComplete();
+                }
+                catch (Exception ex)
+                {
+                    state.PipelineFailed(ex);
+                }
+            });
+
+            return await state.WebSocketTask;
+        }
+
+        private class RequestState : IHttpWebSocketFeature
+        {
+            public IFeatureCollection FeatureCollection { get; private set; }
+            public Task<System.Net.WebSockets.WebSocket> WebSocketTask { get { return _websocketTcs.Task; } }
+
+            private TaskCompletionSource<System.Net.WebSockets.WebSocket> _websocketTcs;
+
+            public RequestState(HttpRequestMessage clientRequest, PathString pathBase, CancellationToken cancellationToken)
+            {
+                _websocketTcs = new TaskCompletionSource<System.Net.WebSockets.WebSocket>();
+
+                // HttpContext
+                FeatureCollection = new FeatureCollection();
+                var httpContext = new DefaultHttpContext(FeatureCollection);
+
+                // Request
+                httpContext.SetFeature<IHttpRequestFeature>(new RequestFeature());
+                var serverRequest = httpContext.Request;
+                serverRequest.Protocol = "HTTP/" + clientRequest.Version.ToString(2);
+                serverRequest.Scheme = clientRequest.RequestUri.Scheme;
```

Comment:
If the user passes in `ws` or `wss` it should be converted to `http` or `https` respectively for the server.